Dmr/django 2.2

dannyroberts opened the pull request:
Opening for tests, don't intend to merge directly. (It contains some unmerged branches, plus a few fixes that I'll cherry pick into their PRs.)

dannyroberts committed:
Upgrade django to 2.2 (run make requirements)

dannyroberts committed:
Fix DatabaseSchemaEditorIfNotExists for Django 2

snopoke commented on corehq/util/django_migrations.py:
```diff
@@ -53,9 +53,13 @@ def add_if_not_exists_raw(string, name):
+    if hasattr(DatabaseSchemaEditor, 'sql_create_varchar_index'):
+        sql_create_varchar_index = add_if_not_exists(DatabaseSchemaEditor.sql_create_varchar_index)
+    if hasattr(DatabaseSchemaEditor, 'sql_create_text_index'):
+        sql_create_text_index = add_if_not_exists(DatabaseSchemaEditor.sql_create_text_index)
```
will these still work without them?

dannyroberts replied:
I had a hunch that they would, because these are internals that were removed and whose calls were replaced within `django`. My hunch is that the calls would be routed through the other properties this overrides.

dannyroberts committed:
[Django 2] Fix icds_reports TestQueryDiffs by making it a db TestCase

dannyroberts committed:
Fix handler403 callback to take exceptions arg

dannyroberts committed:
Restore __hash__ on Subscription

I think even better would be to rename the method now called __eq__
to a normal method (e.g. "is_equivalent_to") that better captures what
it is doing. It doesn't check if it's the same db object like model
normally does, but rather check to see if it is an equivalent
subscription, meaning that two different rows in the db can have
row1 == row2 be True, which is surprising behavior.

dannyroberts committed:
[Django 2] Fix signature of custom handler{404,500}

dannyroberts committed:
[Django 2] Only pass strings to django.utils.http.urlquote

dannyroberts committed:
[Django 2] Fix MultiDBRouter.allow_relation to work during model __init__

The need for this change was introduced by
https://github.com/django/django/commit/a5a2ceeb453b86caaf01687342984f372428beb8
"Added router.allow_relation() calls for assignments between unsaved model instances."

dannyroberts committed:
[Django 2] Update django-nose to 1.4.6

dannyroberts committed:
[Django 2] Disable a broken and now irrelvant part of migrations/0008_deletedblobmeta.py

dannyroberts committed:
Merge branch 'master' into dmr/django-2.2

dannyroberts committed:
[Django 2] Squash last four users migrations

0011 leaves SQLInvitation in a state that is illegal in Django 2,
but that Django 1 was silently accepting, and the only way around it
I could find was to squash the four migrations involved in SQLInvitation,
and just "get it right" in one migration file.

Command and command output were:

```
$ ./manage.py squashmigrations users 0009 0012
Will squash the following migrations:
 - 0009_sqlinvitation
 - 0010_populate_usersinvitation
 - 0011_sqlinvitation_uuid
 - 0012_alter_usersinvitation
Do you wish to proceed? [yN] y
Optimizing...
  Optimized from 7 operations to 1 operations.
Created new squashed migration /Users/danielroberts/dimagi/commcare-hq/corehq/apps/users/migrations/0009_sqlinvitation_squashed_0012_alter_usersinvitation.py
  You should commit this migration but leave the old ones in place;
  the new migration will be used for new installs. Once you are sure
  all instances of the codebase have applied the migrations you squashed,
  you can delete them.
```

orangejenny commented on corehq/apps/users/migrations/0009_sqlinvitation_squashed_0012_alter_usersinvitation.py:
```diff
@@ -0,0 +1,35 @@
+        ('users', '0008_auto_20200129_1852'),
+    ]
+
+    operations = [
```
This is missing the `RunPython` from the old 0010, so the new table isn't going to get populated.

Squashing 0011 and 0012 is fine, but squashing all four means the migration won't work as intended even if the `RunPython` is added, because invitations created while the deploy/migration is running may not be correctly migrated to sql.

This might be fine, because our environments have all had all of these migrations run, and third parties are unlikely to even use invitations, but I wanted to be sure you were aware.

dannyroberts committed:
[Django 2] Use standard url/include instead of URLResolver

The previous change I made attempting to fix was overly complicated
and didn't actually work.

dannyroberts committed:
Use .set() instead of =-assignment on many-to-many field

dannyroberts committed:
[Django 2] Surgically rearrange SQLInvitation db migrations

0011 leaves SQLInvitation in a state that is illegal in Django 2,
but that Django 1 was silently accepting.
Originally we tried squashing the last four migrations
(https://github.com/dimagi/commcare-hq/pull/27114\)
but there was then no reasonable way to sequence the
custom data migration part.

So instead we are surgically merging the 0011 back into
the original migration that created SQLInvitation
in a way that still leaves it workable for
the custom migration code which runs during 0010

dannyroberts committed:
Revert "[Django 2] Squash last four users migrations"

This reverts commit 069aec3d312ed9e2bdf9ca7887c0861af4dbc4c6.

dannyroberts committed:
Back-merge user migration 0012 into 0009 as well

leaving the full SQLInvitation setup happing once correctly
in the first migration (0009) followed by the data migration (0010)
followed by two emtpy migrations (0011, 0012) for backwards compat

dannyroberts committed:
[Django 2] Upgrade jsonfield to a dj 2.2 compatible version

dannyroberts committed:
[Django 2] Update auditcare views to use LoginView and LogoutView

dannyroberts committed:
[Django 2] Allow partitioned tests to interact with all DATABASES
